File: mistral/typescript/src/services/files.ts
/**
 * Files service.
 */

import type { HttpTransport } from '../transport';
import type {
  FileObject,
  FileListResponse,
  FileDeleteResponse,
  FileSignedUrlResponse,
  FileUploadOptions,
} from '../types/files';

/**
 * Files service interface.
 */
export interface FilesService {
  /** Lists all uploaded files. */
  list(): Promise<FileListResponse>;

  /** Retrieves a specific file. */
  retrieve(fileId: string): Promise<FileObject>;

  /** Uploads a file. */
  upload(options: FileUploadOptions): Promise<FileObject>;

  /** Deletes a file. */
  delete(fileId: string): Promise<FileDeleteResponse>;

  /** Gets a signed URL for downloading a file. */
  getSignedUrl(fileId: string): Promise<FileSignedUrlResponse>;
}

/**
 * Default implementation of the files service.
 */
export class DefaultFilesService implements FilesService {
  constructor(private readonly transport: HttpTransport) {}

  async list(): Promise<FileListResponse> {
    const body = await this.transport.get('/v1/files');
    return JSON.parse(body) as FileListResponse;
  }

  async retrieve(fileId: string): Promise<FileObject> {
    const body = await this.transport.get(`/v1/files/${fileId}`);
    return JSON.parse(body) as FileObject;
  }

  async upload(options: FileUploadOptions): Promise<FileObject> {
    const body = await this.transport.uploadFile(
      '/v1/files',
      options.file,
      options.filename,
      options.purpose
    );
    return JSON.parse(body) as FileObject;
  }

  async delete(fileId: string): Promise<FileDeleteResponse> {
    const body = await this.transport.delete(`/v1/files/${fileId}`);
    return JSON.parse(body) as FileDeleteResponse;
  }

  async getSignedUrl(fileId: string): Promise<FileSignedUrlResponse> {
    const body = await this.transport.get(`/v1/files/${fileId}/url`);
    return JSON.parse(body) as FileSignedUrlResponse;
  }
}

/**
 * Creates a files service.
 */
export function createFilesService(transport: HttpTransport): FilesService {
  return new DefaultFilesService(transport);
}

File: mistral/rust/src/services/files.rs
//! Files service.

use async_trait::async_trait;

use crate::errors::MistralError;
use crate::types::files::{
    FileDeleteResponse, FileListResponse, FileObject, FileSignedUrlResponse, FileUploadRequest,
};

/// Files service trait.
#[async_trait]
pub trait FilesService: Send + Sync {
    /// Lists all uploaded files.
    async fn list(&self) -> Result<FileListResponse, MistralError>;

    /// Retrieves a specific file.
    async fn retrieve(&self, file_id: &str) -> Result<FileObject, MistralError>;

    /// Uploads a file.
    async fn upload(&self, request: FileUploadRequest) -> Result<FileObject, MistralError>;

    /// Deletes a file.
    async fn delete(&self, file_id: &str) -> Result<FileDeleteResponse, MistralError>;

    /// Gets a signed URL for downloading a file.
    async fn get_signed_url(&self, file_id: &str) -> Result<FileSignedUrlResponse, MistralError>;
}

/// Default implementation of the files service.
pub struct DefaultFilesService<T> {
    transport: T,
}

impl<T> DefaultFilesService<T> {
    /// Creates a new files service.
    pub fn new(transport: T) -> Self {
        Self { transport }
    }
}

#[async_trait]
impl<T> FilesService for DefaultFilesService<T>
where
    T: crate::transport::HttpTransport + Send + Sync,
{
    async fn list(&self) -> Result<FileListResponse, MistralError> {
        let response = self.transport
            .get("/v1/files")
            .await?;

        serde_json::from_slice(&response)
            .map_err(|e| MistralError::Deserialization {
                message: e.to_string(),
                body: String::from_utf8_lossy(&response).to_string(),
            })
    }

    async fn retrieve(&self, file_id: &str) -> Result<FileObject, MistralError> {
        let path = format!("/v1/files/{}", file_id);
        let response = self.transport
            .get(&path)
            .await?;

        serde_json::from_slice(&response)
            .map_err(|e| MistralError::Deserialization {
                message: e.to_string(),
                body: String::from_utf8_lossy(&response).to_string(),
            })
    }

    async fn upload(&self, request: FileUploadRequest) -> Result<FileObject, MistralError> {
        let response = self.transport
            .post_multipart(
                "/v1/files",
                request.file,
                &request.filename,
                &serde_json::to_string(&request.purpose)
                    .map_err(|e| MistralError::Serialization { message: e.to_string() })?
                    .trim_matches('"'),
            )
            .await?;

        serde_json::from_slice(&response)
            .map_err(|e| MistralError::Deserialization {
                message: e.to_string(),
                body: String::from_utf8_lossy(&response).to_string(),
            })
    }

    async fn delete(&self, file_id: &str) -> Result<FileDeleteResponse, MistralError> {
        let path = format!("/v1/files/{}", file_id);
        let response = self.transport
            .delete(&path)
            .await?;

        serde_json::from_slice(&response)
            .map_err(|e| MistralError::Deserialization {
                message: e.to_string(),
                body: String::from_utf8_lossy(&response).to_string(),
            })
    }

    async fn get_signed_url(&self, file_id: &str) -> Result<FileSignedUrlResponse, MistralError> {
        let path = format!("/v1/files/{}/url", file_id);
        let response = self.transport
            .get(&path)
            .await?;

        serde_json::from_slice(&response)
            .map_err(|e| MistralError::Deserialization {
                message: e.to_string(),
                body: String::from_utf8_lossy(&response).to_string(),
            })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::files::FilePurpose;

    #[test]
    fn test_file_upload_request_creation() {
        let request = FileUploadRequest::new(
            vec![1, 2, 3, 4],
            "test.jsonl",
            FilePurpose::FineTune,
        );

        assert_eq!(request.filename, "test.jsonl");
        assert_eq!(request.file.len(), 4);
        assert_eq!(request.purpose, FilePurpose::FineTune);
    }
}
